sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,346 @@
+            }
+
+            var identifier = Guid.NewGuid().ToString();
+
+            var item = Pool.GetOrAdd(partitionId, id => new PoolItem(partitionId, Connection.CreateTransportProducer(id, RetryPolicy), removeAfterDuration));
+
+            // A race condition at this point may end with CloseAsync called on
+            // the returned PoolItem if it had expired. The probability is very low and
+            // possible exceptions should be handled by the invoking methods.
+
+            if (item.PartitionProducer.IsClosed || !item.ActiveInstances.TryAdd(identifier, 0))
+            {
+                identifier = Guid.NewGuid().ToString();
+                item = Pool.GetOrAdd(partitionId, id => new PoolItem(partitionId, Connection.CreateTransportProducer(id, RetryPolicy), removeAfterDuration));
+                item.ActiveInstances.TryAdd(identifier, 0);
+            }
+
+            item.ExtendRemoveAfter(removeAfterDuration);
+
+            return new PooledProducer(item.PartitionProducer, cleanUp: producer =>
+            {
+                Argument.AssertNotNull(producer, nameof(producer));
+
+                if (Pool.TryGetValue(partitionId, out PoolItem pooledItem))
+                {
+                    pooledItem.ExtendRemoveAfter(removeAfterDuration);
+                }
+
+                // If TryRemove returned false the PoolItem would not be closed deterministically
+                // and the ExpirationTimer callback would eventually remove it from the
+                // Pool leaving to the Garbage Collector the responsability of closing
+                // the TransportProducer and the AMQP link.
+
+                item.ActiveInstances.TryRemove(identifier, out _);
+
+                // The second TryGetValue runs after the extension would have been seen, so it
+                // is intended to be sure that the item wasn't removed in the meantime.
+
+                if (!Pool.TryGetValue(partitionId, out _) && !item.ActiveInstances.Any())
+                {
+                    return producer.CloseAsync(CancellationToken.None);
+                }
+
+                return Task.CompletedTask;
+            });
+        }
+
+        /// <summary>
+        ///   Performs the task needed to clean up resources used by the <see cref="TransportProducerPool" />.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async ValueTask DisposeAsync() => await CloseAsync().ConfigureAwait(false);
+
+        /// <summary>
+        ///   Closes the producers in the pool and performs any cleanup necessary
+        ///   for resources used by the <see cref="TransportProducerPool" />.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task CloseAsync(CancellationToken cancellationToken = default)
```

Comment:
If we're intending to have this disposed, and it's an internal type.... do you think we can get away without a `CloseAsync` method?   We have those hanging out on the clients to make it easier for those working in C# 7.x since they don't have access to the async disposable yet.

Reply:
> We have those hanging out on the clients to make it easier for those working in C# 7.x since they don't have access to the async disposable yet.

I am sure I had been asking myself this question at some point, thank you for flagging that to me.

> If we're intending to have this disposed, and it's an internal type

I might be asking the wrong question here, if we never use ```async using``` coupled with ```TransportProducerPool``` why are we implementing ```IAsyncDisposable```?

Another sticky point I am wondering if you wouldn't mind discussing.

```TransportProducerPool::CloseAsync()``` is closing all the transport producers. An exception triggered closing any of them may get in the way of closing the others.

I wonder if you think we should do anything with that?

Reply:
>> I might be asking the wrong question here, if we never use async using coupled with TransportProducerPool why are we implementing IAsyncDisposable?

On the contrary, I think it **_exactly_** the question that needs to be asked.   Looking over things, you're absolutely correct.  I don't see us doing any disposal of the pool, just the item.   

Based on that, I think the correct answer here is to kill `IAsyncDisposable` in favor of `CloseAsync`

Reply:
> TransportProducerPool::CloseAsync() is closing all the transport producers. An exception triggered closing any of them may get in the way of closing the others.

Good catch.  Yes.  We should try/catch/log and ignore exceptions closing them and remove from the pool into the ether.   The connection timeout will take care of the network responsibilities from the server end and GC should take care of the local resource.